service/matching/workers/registry.go

```diff
@@ -0,0 +1,228 @@
+	"time"
+
+	workerpb "go.temporal.io/api/worker/v1"
+)
+
+// entry wraps a WorkerHeartbeat along with its namespace and eviction metadata.
+type entry struct {
+	namespace string
+	hb        *workerpb.WorkerHeartbeat
+	lastSeen  time.Time
+	elem      *list.Element
+}
+
+// bucket holds part of the keyspace: a map from namespace → (map of instanceKey → entry),
+// plus a recency list for eviction.
+type bucket struct {
+	mu         sync.Mutex
+	namespaces map[string]map[string]*entry
+	order      *list.List // front = oldest, back = newest
+}
+
+func newBucket() *bucket {
+	return &bucket{
+		namespaces: make(map[string]map[string]*entry),
+		order:      list.New(),
+	}
+}
+
+// update inserts or refreshes a WorkerHeartbeat under the given namespace.
+// Returns true if a brand-new entry was created.
+func (b *bucket) update(namespace string, hb *workerpb.WorkerHeartbeat) bool {
+	now := time.Now()
+	key := hb.WorkerInstanceKey
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	mp, ok := b.namespaces[namespace]
+	if !ok {
+		mp = make(map[string]*entry)
+		b.namespaces[namespace] = mp
+	}
+
+	if e, exists := mp[key]; exists {
+		e.hb = hb
+		e.lastSeen = now
+		b.order.MoveToBack(e.elem)
+		return false
+	}
+
+	e := &entry{
+		namespace: namespace,
+		hb:        hb,
+		lastSeen:  now,
+	}
+	e.elem = b.order.PushBack(e)
+	mp[key] = e
+	return true
+}
+
+// evictByTTL removes entries older than expireBefore from this bucket.
+// Returns the number of entries removed.
+func (b *bucket) evictByTTL(expireBefore time.Time) int {
+	removed := 0
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	for {
+		front := b.order.Front()
+		if front == nil {
+			break
+		}
+		e := front.Value.(*entry) //nolint:revive
+		if !e.lastSeen.Before(expireBefore) {
+			break
+		}
+		b.order.Remove(front)
+		delete(b.namespaces[e.namespace], e.hb.WorkerInstanceKey)
+		removed++
+	}
+	return removed
+}
+
+// Registry contains all worker heartbeats.
+// It partitions the keyspace into buckets and enforces TTL and capacity.
+// Eviction runs in the background.
+type Registry struct {
+	buckets     []*bucket
+	maxItems    int64
+	TTL         time.Duration
+	MinEvictAge time.Duration
+	ticker      *time.Ticker
+	hashFn      func(string) uint32
+	total       atomic.Int64 // atomic counter of total entries
+	quit        chan struct{}
+}
+
+// NewRegistry creates a workers heartbeat registry with the given parameters.
+func NewRegistry(
+	numBuckets int,
+	ttl time.Duration,
+	minEvictAge time.Duration,
+	maxItems int64,
+	evictionInterval time.Duration,
+) *Registry {
+	m := &Registry{
+		buckets:     make([]*bucket, numBuckets),
+		maxItems:    maxItems,
+		TTL:         ttl,
+		MinEvictAge: minEvictAge,
+		ticker:      time.NewTicker(evictionInterval),
+		hashFn: func(s string) uint32 {
+			h := fnv.New32a()
+			h.Write([]byte(s)) //nolint:revive
+			return h.Sum32()
+		},
+		quit: make(chan struct{}),
+	}
+	for i := range m.buckets {
+		m.buckets[i] = newBucket()
+	}
+	go m.evictLoop()
```

Comment:
In general most objects with background goroutines support `Start()` and `Stop()` methods and only create goroutines in Start, not the constructor. that's more predictable and makes it easy to integrate with fx lifecycle.